packages/core/src/node/messaging/http-websocket-adapter.ts

```diff
@@ -0,0 +1,83 @@
+/********************************************************************************
+ * Copyright (C) 2021 TypeFox and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+/* eslint-disable @typescript-eslint/no-explicit-any */
+
+import { injectable } from 'inversify';
+import * as ws from 'ws';
+
+export const HttpWebsocketAdapterFactory = Symbol('HttpWebsocketAdapterFactory');
+
+@injectable()
+export class HttpWebsocketAdapter {
+
+    readyState: number = ws.OPEN;
+    alive: boolean = true;
+
+    protected pendingTimeout: NodeJS.Timeout | undefined;
+    protected pendingMessages: unknown[] = [];
+    protected pendingMessagesResolve?: (messages: unknown[]) => void;
+
+    getPendingMessages(): Promise<unknown[]> {
+        this.alive = true;
+        return new Promise(resolve => {
```

Comment:
I might use a `Deferred` here, rather than creating a `Promise` and conditionally assigning its `render` callback to the class. The effect would be the same, but I think the intent might be clearer.

Reply:
Thanks for the hint in the `Deferred` direction. The code looks way cleaner now. I had to add state to the `Deferred` class though, to find out whether the promise is already resolved. Let me now if you find a better solution to that.